{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4d893960-0caa-4f2e-a560-5f17d09058d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0] <h1>\n",
      "  <picture>\n",
      "    <source media=\"(prefers-color-scheme: dark)\" srcset=\"docs/images/nf-core-pairgenomealign_logo_dark.png\">\n",
      "    <img alt=\"nf-core/pairgenomealign\" src=\"docs/images/nf-core-pairgenomealign_logo_light.png\">\n",
      "  </picture>\n",
      "</h1>\n",
      "\n",
      "[![GitHub Actions CI Status](https://github.com/nf-core/pairgenomealign/actions/workflows/ci.yml/badge.svg)](https://github.com/nf-core/pairgenomealign/actions/workflows/ci.yml)\n",
      "[![GitHub Actions Linting Status](https://github.com/nf-core/pairgenomealign/actions/workflows/linting.yml/badge.svg)](https://github.com/nf-core/pairgenomealign/actions/workflows/linting.yml)[![AWS CI](https://img.shields.io/badge/CI%20tests-full%20size-FF9900?labelColor=000000&logo=Amazon%20AWS)](https://nf-co.re/pairgenomealign/results)[![Cite with Zenodo](http://img.shields.io/badge/DOI-10.5281/zenodo.13910535-1073c8?labelColor=000000)](https://doi.org/10.5281/zenodo.13910535)\n",
      "[![nf-test](https://img.shields.io/badge/unit_tests-nf--test-337ab7.svg)](https://www.nf-test.com)\n",
      "\n",
      "[![Nextflow](https://img.shields.io/badge/nextflow%20DSL2-%E2%89%A524.10.1-23aa62.svg)](https://www.nextflow.io/)\n",
      "[![run with conda](http://img.shields.io/badge/run%20with-conda-3EB049?labelColor=000000&logo=anaconda)](https://docs.conda.io/en/latest/)\n",
      "[![run with docker](https://img.shields.io/badge/run%20with-docker-0db7ed?labelColor=000000&logo=docker)](https://www.docker.com/)\n",
      "[![run with singularity](https://img.shields.io/badge/run%20with-singularity-1d355c.svg?labelColor=000000)](https://sylabs.io/docs/)\n",
      "[![Launch on Seqera Platform](https://img.shields.io/badge/Launch%20%F0%9F%9A%80-Seqera%20Platform-%234256e7)](https://cloud.seqera.io/launch?pipeline=https://github.com/nf-core/pairgenomealign)\n",
      "\n",
      "[![Get help on Slack](http://img.shields.io/badge/slack-nf--core%20%23pairgenomealign-4A154B?labelColor=000000&logo=slack)](https://nfcore.slack.com/channels/pairgenomealign)[![Follow on Twitter](http://img.shields.io/badge/twitter-%40nf__core-1DA1F2?labelColor=000000&logo=twitter)](https://twitter.com/nf_core)[![Follow on Mastodon](https://img.shields.io/badge/mastodon-nf__core-6364ff?labelColor=FFFFFF&logo=mastodon)](https://mstdn.science/@nf_core)[![Watch on YouTube](http://img.shields.io/badge/youtube-nf--core-FF0000?labelColor=000000&logo=youtube)](https://www.youtube.com/c/nf-core)\n",
      "\n",
      "\n",
      "\n",
      "[1]  Introduction\n",
      "\n",
      "**nf-core/pairgenomealign** is a bioinformatics pipeline that aligns one or more _query_ genomes to a _target_ genome, and plots pairwise representations.\n",
      "\n",
      "![Tubemap workflow summary](docs/images/pairgenomealign-tubemap.png \"Tubemap workflow summary\")\n",
      "\n",
      "The main steps of the pipeline are:\n",
      "\n",
      "1. Genome QC ([`assembly-scan`](https://github.com/rpetit3/assembly-scan)).\n",
      "2. Genome indexing ([`lastdb`](https://gitlab.com/mcfrith/last/-/blob/main/doc/lastdb.rst)).\n",
      "3. Genome pairwise alignments ([`lastal`](https://gitlab.com/mcfrith/last/-/blob/main/doc/lastal.rst)).\n",
      "4. Alignment plotting ([`last-dotplot`](https://gitlab.com/mcfrith/last/-/blob/main/doc/last-dotplot.rst)).\n",
      "5. Alignment export to various formats with [`maf-convert`](https://gitlab.com/mcfrith/last/-/blob/main/doc/maf-convert.rst), plus [`Samtools`](https://www.htslib.org/) for SAM/BAM/CRAM.\n",
      "\n",
      "The pipeline can generate four kinds of outputs, called _many-to-many_, _many-to-one_, _one-to-many_ and _one-to-one_, depending on whether sequences of one genome are allowed match the other genome multiple times or not.\n",
      "\n",
      "These alignments are output in [MAF](https://genome.ucsc.edu/FAQ/FAQformat.html#format5) format, and optional line plot representations are output in PNG format.\n",
      "\n",
      "\n",
      "\n",
      "[2]  Usage\n",
      "\n",
      "> [!NOTE]\n",
      "> If you are new to Nextflow and nf-core, please refer to [this page](https://nf-co.re/docs/usage/installation) on how to set-up Nextflow. Make sure to [test your setup](https://nf-co.re/docs/usage/introduction#how-to-run-a-pipeline) with `-profile test` before running the workflow on actual data.\n",
      "\n",
      "First, prepare a samplesheet with your input data that looks as follows:\n",
      "\n",
      "`samplesheet.csv`:\n",
      "\n",
      "```csv\n",
      "sample,fasta\n",
      "query_1,path-to-query-genome-file-one.fasta\n",
      "query_2,path-to-query-genome-file-two.fasta\n",
      "```\n",
      "\n",
      "Each row represents a fasta file, this can also contain multiple rows to accomodate multiple query genomes in fasta format.\n",
      "\n",
      "Now, you can run the pipeline using:\n",
      "\n",
      "```bash\n",
      "nextflow run nf-core/pairgenomealign \\\n",
      "   -profile <docker/singularity/.../institute> \\\n",
      "   --target sequencefile.fa \\\n",
      "   --input samplesheet.csv \\\n",
      "   --outdir <OUTDIR>\n",
      "```\n",
      "\n",
      "> [!WARNING]\n",
      "> Please provide pipeline parameters via the CLI or Nextflow `-params-file` option. Custom config files including those provided by the `-c` Nextflow option can be used to provide any configuration _**except for parameters**_; see [docs](https://nf-co.re/docs/usage/getting_started/configuration#custom-configuration-files).\n",
      "\n",
      "For more details and further functionality, please refer to the [usage documentation](https://nf-co.re/pairgenomealign/usage) and the [parameter documentation](https://nf-co.re/pairgenomealign/parameters).\n",
      "\n",
      "\n",
      "\n",
      "[3]  Pipeline output\n",
      "\n",
      "To see the results of an example test run with a full size dataset refer to the [results](https://nf-co.re/pairgenomealign/results) tab on the nf-core website pipeline page.\n",
      "For more details about the output files and reports, please refer to the\n",
      "[output documentation](https://nf-co.re/pairgenomealign/output).\n",
      "\n",
      "\n",
      "\n",
      "[4]  Credits\n",
      "\n",
      "`nf-core/pairgenomealign` was originally written by [charles-plessy](https://github.com/charles-plessy); the original versions are available at <https://github.com/oist/plessy_pairwiseGenomeComparison>.\n",
      "\n",
      "We thank the following people for their extensive assistance in the development of this pipeline:\n",
      "\n",
      "- [Mahdi Mohammed](https://github.com/U13bs1125) ported the original pipeline to _nf-core_ template 2.14.x.\n",
      "- [Martin Frith](https://github.com/mcfrith/), the author of LAST, gave us extensive feedback and advices.\n",
      "- [Michael Mansfield](https://github.com/mjmansfi) tested the pipeline and provided critical comments.\n",
      "- [Aleksandra Bliznina](https://github.com/aleksandrabliznina) contributed to the creation of the initial `last/*` modules.\n",
      "- [Jiashun Miao](https://github.com/miaojiashun) and [Huyen Pham](https://github.com/ngochuyenpham) tested the pipeline on vertebrate genomes.\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from typing import List\n",
    "\n",
    "def split_into_chunks(doc_file: str) -> List[str]:\n",
    "    with open(doc_file, 'r',encoding='utf-8') as file:\n",
    "        content = file.read()\n",
    "    return [chunk for chunk in content.split(\"##\")]\n",
    "\n",
    "# 主程序部分\n",
    "chunks = split_into_chunks(\"README.md\")\n",
    "\n",
    "for i, chunk in enumerate(chunks[:5]):\n",
    "    print(f\"[{i}] {chunk}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ce827421-54bf-467f-9364-85ec4a8e0ca5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "384\n",
      "[0.011573436670005322, 0.025136202573776245, -0.03670184686779976, 0.05932488292455673, -0.0071490490809082985, -0.04119422286748886, 0.0770873948931694, 0.03744256868958473, 0.01244899071753025, -0.006117628887295723, 0.017034275457262993, -0.07701538503170013, -0.00039416426443494856, 0.027909062802791595, -0.015989158302545547, -0.06827527284622192, 0.008884645998477936, -0.020280703902244568, -0.08035995811223984, -0.013074046932160854, -0.04110001400113106, -0.025898080319166183, -0.0265386700630188, 0.03305228799581528, -0.022079195827245712, 0.021046103909611702, -0.05792200192809105, 0.03294876217842102, 0.02970738522708416, -0.06224840506911278, 0.038788024336099625, 0.03199068829417229, 0.015330815687775612, 0.0453069731593132, 0.05314944311976433, 0.013360676355659962, 0.041224926710128784, 0.028142910450696945, 0.019398434087634087, -0.0032523232512176037, -0.0036123408935964108, -0.14286024868488312, 0.0380711704492569, -0.010916205123066902, 0.02609400637447834, 0.04136994853615761, -0.016015755012631416, 0.05356009677052498, -0.056859441101551056, 0.012246801517903805, -0.03499660640954971, -0.03975418210029602, -0.04614299163222313, -0.03911236673593521, -0.018003640696406364, 0.021634286269545555, -0.006461341865360737, -0.026569491252303123, 0.04872880131006241, 0.043400608003139496, 0.04620008543133736, -0.034479010850191116, -0.024219267070293427, 0.055649109184741974, 0.024624425917863846, 0.01948554255068302, 0.01135282963514328, -0.02576228231191635, -0.032244496047496796, -0.032652173191308975, -0.01459591370075941, 0.014690334908664227, 0.01030451338738203, 0.07407969981431961, 0.08006177097558975, -0.0041435472667217255, 0.0027142451144754887, -0.07674671709537506, 0.05881531164050102, -0.004839183297008276, -0.08736330270767212, -0.046461936086416245, -0.03962543234229088, 0.05249390751123428, 0.02480141632258892, 0.08080153167247772, 0.11981367319822311, 0.03177144378423691, -0.11397235840559006, 0.010673847049474716, 0.019723786041140556, 0.03871829807758331, -0.03454233705997467, -0.006243126932531595, -0.04255075380206108, 0.0226285383105278, 0.008293666876852512, -0.019443608820438385, -0.023251060396432877, 0.24533765017986298, 0.04944216087460518, 0.029147351160645485, -0.005753715988248587, -0.02041955292224884, -0.07761149108409882, -0.042979251593351364, 0.0004302800807636231, -0.06692595034837723, 0.0707603469491005, 0.006507022771984339, -0.05844533443450928, -0.00505733210593462, 0.030943766236305237, 0.027579613029956818, 0.027111228555440903, -0.06818616390228271, -0.0657687708735466, 0.05907768756151199, -0.03189412131905556, 0.0406060554087162, 0.05710183084011078, 0.0069035436026751995, 0.006207626312971115, -0.013271474279463291, 0.03124042972922325, -0.04029698669910431, 0.070852130651474, -4.8462737429564014e-33, 0.02193438448011875, -0.10270480811595917, 0.05621321126818657, 0.09758371859788895, -0.05268485099077225, 0.019097719341516495, -0.01050743367522955, 0.07035725563764572, -0.009123295545578003, 0.05957067012786865, 0.009338634088635445, -0.01552300900220871, -0.023274077102541924, 0.023976817727088928, 0.1020885482430458, 0.09320350736379623, -0.03313610330224037, 0.011721611954271793, -0.06560058146715164, 0.03146481141448021, -0.02380792424082756, -0.048739075660705566, 0.012257893569767475, -0.04001418873667717, -0.07650820165872574, -0.0355532206594944, -0.0013622069964185357, -0.01642921008169651, 0.01688247174024582, -0.0003748929884750396, -0.05272497609257698, 0.029880939051508904, -0.07291102409362793, 0.0691225528717041, -0.017607947811484337, -0.005510292947292328, 0.012954636476933956, -0.022701941430568695, 0.02674126997590065, -0.025845780968666077, -0.0402020700275898, -0.013495084829628468, 0.0007997856009751558, 0.028601309284567833, 0.031942546367645264, -0.03160711005330086, -0.029549865052103996, -0.02025127038359642, 0.04817778989672661, -0.001307626604102552, -0.013747242279350758, 0.020030170679092407, -0.06870564818382263, -0.021960807964205742, -0.031332165002822876, 0.04928040876984596, 0.012094746343791485, -0.058868054300546646, -0.02646571211516857, 0.05988997220993042, 0.06764886528253555, 0.03401569649577141, -0.05288434028625488, 0.05971936509013176, -0.0254825372248888, -0.020729975774884224, -0.053825538605451584, -0.09741067886352539, 0.04792823642492294, 0.05239955708384514, -0.023259682580828667, -0.06907136738300323, 0.016656821593642235, 0.028476404026150703, -0.02920488826930523, -0.035486698150634766, -0.012644192203879356, 0.07333981990814209, -0.019434962421655655, -0.06327889114618301, 0.09606637805700302, -0.07743865251541138, 0.01593954674899578, -0.04480079561471939, 0.016302715986967087, -0.0007480642525479198, -0.0087027158588171, -0.09881404787302017, 0.005742848850786686, -0.07192353904247284, -0.03267769515514374, 0.019894618541002274, 0.0038597467355430126, -0.025551123544573784, 0.08238352090120316, 4.08626411524599e-33, -0.02948814444243908, 0.0255514457821846, -0.051060862839221954, 0.15531232953071594, 0.05231140926480293, -0.034548208117485046, 0.13314689695835114, -0.01920968107879162, -0.05976710096001625, 0.12289653718471527, 0.010202900506556034, -0.0496731661260128, 0.058467235416173935, 0.01273270696401596, -0.016586363315582275, 0.012795469723641872, 0.04575824365019798, -0.06982145458459854, -0.0485241673886776, -0.004963469225913286, -0.09043657779693604, 0.06992115825414658, 0.009388985112309456, -0.0067446171306073666, -0.10609240084886551, 0.031094854697585106, 0.049425870180130005, -0.044871050864458084, -0.007371742743998766, -0.03356107696890831, 0.07605849206447601, 0.007239074911922216, -0.042201507836580276, 0.07079145312309265, 0.04747209697961807, 0.020783035084605217, 0.153310626745224, -0.008394013158977032, -0.02588072419166565, 0.06079910695552826, 0.06681554764509201, 0.06472287327051163, 0.04982012137770653, 0.08878505975008011, -0.03294116258621216, 0.07035742700099945, 0.017195379361510277, -0.0301852747797966, 0.03854382038116455, 0.0484696663916111, -0.06051015481352806, 0.03053242526948452, 0.015603824518620968, -0.030421435832977295, -0.009440488182008266, -0.04105142503976822, -0.06789776682853699, 0.01019968744367361, -0.025656607002019882, 0.02171560376882553, -0.06997860968112946, 0.09247466921806335, -0.03571975603699684, 0.07013790309429169, -0.06342041492462158, -0.032940249890089035, -0.04619637131690979, 0.05414016172289848, 0.05172332003712654, 0.04292222484946251, 0.013475080020725727, 0.016596594825387, -0.04410724714398384, -0.019720006734132767, 0.03620153293013573, -0.019661419093608856, -0.11567910760641098, 0.0059549640864133835, 0.004566865973174572, -0.04494292661547661, -0.06840216368436813, -0.08530456572771072, -0.07095210999250412, 0.08038385212421417, -0.05798295885324478, 0.0578271709382534, 0.05022653192281723, 0.0594230480492115, -0.0365564301609993, 0.009269560687243938, 0.05252370610833168, 0.027989543974399567, -0.03336906060576439, -0.05078492313623428, -0.012864798307418823, -1.429785445594689e-08, -0.04052524268627167, -0.08579086512327194, 0.04516833648085594, 0.021677004173398018, -0.02233848348259926, 0.0122076366096735, -0.03248917683959007, -0.016953006386756897, -0.02717105858027935, 0.006002947688102722, 0.04027610644698143, 0.02696269564330578, -0.03562461584806442, 0.07408861070871353, 0.03237410634756088, -0.09056802093982697, -0.031741656363010406, 0.04092533513903618, -0.009955926798284054, 0.030688324943184853, -0.07691401988267899, 0.041584499180316925, 0.00019602905376814306, 0.062776580452919, -0.036090560257434845, 0.04884400963783264, 0.05422696843743324, 0.12662003934383392, -0.003848654218018055, 0.0008294829167425632, 0.06961393356323242, 0.04400504752993584, -0.03208102658390999, -0.08523829281330109, 0.01376986876130104, 0.022801825776696205, -0.002847283147275448, -0.00678521441295743, 0.037587955594062805, 0.03527689725160599, -0.06678413599729538, 0.021526463329792023, 0.0375266931951046, -0.04542563855648041, -0.05103161931037903, -0.06799472123384476, -0.030867066234350204, -0.03639037534594536, -0.014874949119985104, -0.09368289262056351, -0.031577371060848236, 0.01024177297949791, 0.015077258460223675, -0.0023830789141356945, 0.024135582149028778, -0.013285377994179726, 0.006583834532648325, 0.024441532790660858, -0.13713593780994415, 0.06391434371471405, 0.19671842455863953, -0.006029724143445492, 0.05319402739405632, -0.05522599443793297]\n"
     ]
    }
   ],
   "source": [
    "from sentence_transformers import SentenceTransformer\n",
    "from typing import List\n",
    "\n",
    "embedding_model = SentenceTransformer(\"all-MiniLM-L6-v2\")\n",
    "\n",
    "def embed_chunk(chunk: str) -> List[float]:\n",
    "    embedding = embedding_model.encode(chunk)\n",
    "    return embedding.tolist()\n",
    "\n",
    "test_embedding = embed_chunk(\"test\")\n",
    "print(len(test_embedding))\n",
    "print(test_embedding)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3ee8886c-0d36-4447-b75f-6eb2c91c1a69",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "72\n",
      "[-0.09221261739730835, 0.018330078572034836, -0.014622059650719166, 0.02387947589159012, 0.04229506105184555, 0.03056703321635723, -0.013751836493611336, -0.03502431884407997, 0.029271598905324936, -0.015627561137080193, 0.0398603230714798, 0.06536902487277985, 0.02458924986422062, -0.042360153049230576, -0.0004565980634652078, 0.0691816583275795, -0.09958221763372421, 0.03261108323931694, -0.0450553372502327, -0.04892238602042198, -0.011267300695180893, -0.06656229496002197, -0.013047306798398495, -0.022809557616710663, -0.01983414962887764, 0.02300063706934452, -0.031199226155877113, -0.0029408191330730915, 0.01665005460381508, -0.07034137099981308, -0.036975931376218796, 0.06463643163442612, 0.035039402544498444, 0.03786580264568329, -0.02161167562007904, 0.05056149885058403, 0.060499317944049835, -0.030701272189617157, -0.01798071898519993, -0.029251625761389732, 0.005497481673955917, -0.060153476893901825, 0.056236788630485535, -0.04040248319506645, 0.02569899894297123, 0.10685986280441284, 0.00851787906140089, -0.025411352515220642, 0.00028341164579614997, -0.021494897082448006, -0.007527649402618408, -0.06972865015268326, -0.024324025958776474, -0.011469176970422268, 0.07349451631307602, 0.0588875412940979, -0.020619459450244904, -0.11931183189153671, 0.026320720091462135, 0.03648310527205467, 0.09503220021724701, 0.02164505422115326, -0.02112896926701069, 0.02522830292582512, -0.015874257311224937, 0.0646442174911499, 0.024151509627699852, -0.04140082746744156, -0.04162610322237015, -0.04651911184191704, -0.041128579527139664, -0.015378451906144619, -0.06900437921285629, -0.04221157729625702, -0.025205183774232864, 0.04996704310178757, 0.01833341084420681, -0.0158835556358099, -0.020858444273471832, -0.14607058465480804, -0.04718592017889023, 0.03798038139939308, 0.03462883085012436, 0.08874977380037308, 0.1477709412574768, 0.06486035138368607, -0.033537451177835464, -0.0320466123521328, 0.06228784844279289, 0.008407269604504108, -0.024881934747099876, 0.027144212275743484, -0.02424856461584568, 0.012853468768298626, -0.029726553708314896, -0.009064489044249058, -0.02943565510213375, 0.03541335090994835, -0.045409079641103745, 0.1201392263174057, -0.010682012885808945, -0.01936972327530384, 0.06970542669296265, -0.04751087352633476, 0.04675016179680824, 0.013073910027742386, 0.030539633706212044, 0.025771643966436386, 0.03612799569964409, 0.04425033554434776, -0.017266280949115753, -0.04925402253866196, -0.04757004231214523, -0.011636056005954742, -0.08009497076272964, 0.04297824576497078, 0.028610959649086, 0.05834619700908661, 0.05716564506292343, -0.04045053571462631, 0.05482561141252518, -0.008436662144958973, -0.12082560360431671, -0.06166573986411095, -0.07603916525840759, -0.016911135986447334, -0.008887668140232563, 6.170352783774534e-33, 0.0745452344417572, 0.05035620182752609, -0.006294112652540207, 0.034709811210632324, 0.09586521238088608, 0.017640825361013412, -0.050969164818525314, -0.2097315490245819, -0.09665475785732269, 0.02126062661409378, 0.04128115251660347, 0.08710600435733795, -0.04527772217988968, 0.013524850830435753, -0.0756872296333313, -0.11716161668300629, -0.00844473298639059, -0.016452481970191002, -0.004292208235710859, 0.008135562762618065, 0.02604111097753048, -0.007622861303389072, -0.10903953015804291, 0.019159989431500435, -0.015669550746679306, 0.03936940059065819, 0.05785525217652321, 0.04478093236684799, -0.029309045523405075, -0.024995671585202217, 0.03657568618655205, 0.013077120296657085, 0.08078393340110779, 0.004568080417811871, -0.0076963575556874275, 6.511920946650207e-05, 0.010737880133092403, -0.049516309052705765, 0.014436235651373863, -0.053616736084222794, 0.006191400811076164, 0.020771993324160576, -0.036622341722249985, -0.046925198286771774, 0.016378700733184814, 0.00018918013665825129, 0.021210838109254837, 0.04066259786486626, 0.10584897547960281, -0.024903951212763786, 0.012276170775294304, 0.0413363054394722, 0.00902616884559393, -0.062106527388095856, -0.061130691319704056, -0.008874528110027313, -0.025084299966692924, 0.07418844848871231, -0.023827670142054558, -0.0025155898183584213, -0.05160604044795036, 0.0009861162398010492, 0.029075253754854202, -0.057995300740003586, 0.036010779440402985, 0.054639335721731186, -0.06986609846353531, 0.01591903157532215, 0.0017274935962632298, -0.09514405578374863, -0.014047465287148952, 0.07103145867586136, 0.057957254350185394, -0.017110371962189674, -0.0007627783925272524, -0.0479552261531353, 0.0247031357139349, 0.006259606685489416, 0.01850232109427452, -0.02074035257101059, -0.13012105226516724, 0.04110296070575714, -0.019816696643829346, -0.01546467561274767, 0.014536844566464424, -0.03674181178212166, 0.034614160656929016, 0.025026259943842888, -0.12709838151931763, 0.030824797227978706, 0.025626743212342262, 0.030515898019075394, 0.04130160063505173, 0.01607988215982914, 0.007597243878990412, -6.231873015308788e-33, 0.0073600769974291325, -0.0020706942304968834, 0.027099721133708954, 0.05993697792291641, 0.06794653087854385, 0.04429546743631363, 0.05729203671216965, 0.010309072211384773, -0.006691805552691221, 0.003536094445735216, 0.13836410641670227, -0.003220983315259218, -0.11184388399124146, -0.04339153692126274, 0.051625557243824005, -0.05398813635110855, 0.02074253186583519, 0.14312945306301117, -0.029008308425545692, 0.02845403552055359, 0.060408301651477814, 0.07942502945661545, -0.029690848663449287, 0.06877412647008896, -0.08914169669151306, 0.07033943384885788, 0.049529992043972015, 0.021728772670030594, -0.0631837397813797, -0.0433085635304451, -0.0016769204521551728, -0.022490154951810837, -0.12638014554977417, 0.009646270424127579, 0.07655094563961029, -0.04091412201523781, -0.04352835193276405, 0.03605327010154724, -0.03678730130195618, 0.007641850505024195, 0.017229661345481873, 0.030485864728689194, -0.04303424432873726, 0.0324089378118515, -0.010458809323608875, 0.020775925368070602, 0.019386453554034233, -0.08910898864269257, -0.056296732276678085, -0.018437175080180168, 0.005842864513397217, -0.020969951525330544, -0.06471742689609528, 0.022507082670927048, -0.05953136458992958, 0.07040243595838547, 0.029056675732135773, 0.037071794271469116, -0.035440344363451004, 0.04558788612484932, 0.011847768910229206, 0.014176832512021065, -0.0647164136171341, -0.0016467799432575703, 0.01941375993192196, -0.04416993632912636, -0.02551443688571453, 0.04890332370996475, 0.003372016828507185, 0.0260462686419487, 0.02779589779675007, 0.048661548644304276, 0.04684973880648613, -0.0598594956099987, 0.059077486395835876, -0.16240963339805603, 0.09406629204750061, 0.004875989630818367, -0.0030775831546634436, 0.0037974012084305286, -0.06953234225511551, 0.016002455726265907, -0.007024142425507307, 0.0792958214879036, -0.03204047307372093, 0.0639193207025528, -0.0029406773392111063, 0.0878366082906723, 0.05668063834309578, -0.0051554967649281025, -0.09303201735019684, 0.022471610456705093, 0.07952943444252014, 0.02455904148519039, -0.014064394868910313, -5.0516828054014695e-08, -0.04358663409948349, -0.011316503398120403, -0.009492112323641777, -0.11536158621311188, -0.01998027414083481, 0.015361173078417778, -0.010523457080125809, -0.024065107107162476, -0.041876405477523804, -0.0695694237947464, 0.041287802159786224, 0.03872049227356911, -0.07010918855667114, -0.07828526943922043, -0.006603749003261328, -0.06098370999097824, -0.025617405772209167, 0.08100207149982452, -0.06113100051879883, -0.03865492343902588, -0.05801181495189667, -0.03290648013353348, 0.021248528733849525, 0.08251875638961792, -0.08759699016809464, -0.036214254796504974, 0.05781707167625427, 0.054377611726522446, 0.030010977759957314, 0.008630895055830479, -0.013149366714060307, 0.05185113474726677, 0.06564152985811234, -0.07747452706098557, -0.04057300463318825, -0.009069967083632946, -0.04454715549945831, -0.01690472476184368, 0.03416985273361206, 0.0318438746035099, -0.005269855726510286, -0.054702214896678925, -0.005666567012667656, 0.04230467230081558, 0.008926617912948132, 0.029270239174365997, 0.03124670870602131, -0.028569666668772697, -0.049841444939374924, -0.006186927203088999, 0.04794958606362343, -0.016812356188893318, 0.0005495123332366347, 0.05411752313375473, -0.06398475170135498, -0.04157523810863495, 0.03504560515284538, 0.047197241336107254, 0.09991028904914856, 0.04789594188332558, 0.008114434778690338, -0.08587347716093063, -0.04133513942360878, 0.05039304122328758]\n"
     ]
    }
   ],
   "source": [
    "embeddings = [embed_chunk(chunk) for chunk in chunks]\n",
    "print(len(embeddings))\n",
    "print(embeddings[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6d8c8e86-6b56-4b3e-a836-499fcf317e79",
   "metadata": {},
   "outputs": [],
   "source": [
    "import chromadb\n",
    "\n",
    "chromadb_client = chromadb.EphemeralClient()\n",
    "chromadb_collection = chromadb_client.get_or_create_collection(name=\"default\")\n",
    "\n",
    "def save_embeddings(chunks:List[str], embeddings:List[List[float]]) -> None:\n",
    "    ids = [str(i) for i in range(len(chunks))]\n",
    "    chromadb_collection.add(\n",
    "        documents=chunks,\n",
    "        embeddings=embeddings,\n",
    "        ids=ids\n",
    "    )\n",
    "\n",
    "save_embeddings(chunks,embeddings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d9e183c3-7023-4fdd-80ec-9dcefe332c7c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0]  Contributions and Support\n",
      "\n",
      "If you would like to contribute to this pipeline, please see the [contributing guidelines](.github/CONTRIBUTING.md).\n",
      "\n",
      "For further information or help, don't hesitate to get in touch on the [Slack `#metatdenovo` channel](https://nfcore.slack.com/channels/metatdenovo) (you can join with [this invite](https://nf-co.re/join/slack)).\n",
      "\n",
      "\n",
      "\n",
      "[1]  Credits\n",
      "\n",
      "nf-core/metatdenovo was originally written by Danilo Di Leo (@danilodileo), Emelie Nilsson (@emnilsson) & Daniel Lundin (@erikrikarddaniel).\n",
      "\n",
      "\n",
      "\n",
      "[2]  Introduction\n",
      "\n",
      "**nf-core/metatdenovo** is a bioinformatics best-practice analysis pipeline for assembly and annotation of metatranscriptomic and metagenomic data from prokaryotes, eukaryotes or viruses.\n",
      "\n",
      "On release, automated continuous integration tests run the pipeline on a full-sized dataset on the AWS cloud infrastructure. This ensures that the pipeline runs on AWS, has sensible resource allocation defaults set to run on real-world datasets, and permits the persistent storage of results to benchmark between pipeline releases and other analysis sources. The results obtained from the full-sized test can be viewed on the [nf-core website](https://nf-co.re/metatdenovo/results).\n",
      "\n",
      "\n",
      "\n",
      "[3]  Pipeline output\n",
      "\n",
      "To see the results of an example test run with a full size dataset refer to the [results](https://nf-co.re/metatdenovo/results) tab on the nf-core website pipeline page.\n",
      "For more details about the output files and reports, please refer to the\n",
      "[output documentation](https://nf-co.re/metatdenovo/output).\n",
      "\n",
      "> [!NOTE]\n",
      "> Tables in `summary_tables` directory under the output directory are made especially for further analysis in tools like R or Python.\n",
      "\n",
      "\n",
      "\n",
      "[4]  Usage\n",
      "\n",
      "![nf-core/metatdenovo metro map](docs/images/metat-metromap.png)\n",
      "\n",
      "1. Read QC ([`FastQC`](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/))\n",
      "2. Present QC for raw reads ([`MultiQC`](http://multiqc.info/))\n",
      "3. Quality trimming and adapter removal for raw reads ([`Trim Galore!`](https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/))\n",
      "4. Optional: Filter sequences with [`BBduk`](https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbduk-guide/)\n",
      "5. Optional: Normalize the sequencing depth with [`BBnorm`](https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbnorm-guide/)\n",
      "6. Merge trimmed, pair-end reads ([`Seqtk`](https://github.com/lh3/seqtk))\n",
      "7. Choice of de novo assembly programs:\n",
      "   1. [`RNAspades`](https://cab.spbu.ru/software/rnaspades/) suggested for Eukaryote de novo assembly\n",
      "   2. [`Megahit`](https://github.com/voutcn/megahit) suggested for Prokaryote de novo assembly\n",
      "8. Choice of orf caller:\n",
      "   1. [`TransDecoder`](https://github.com/TransDecoder/TransDecoder) suggested for Eukaryotes\n",
      "   2. [`Prokka`](https://github.com/tseemann/prokka) suggested for Prokaryotes\n",
      "   3. [`Prodigal`](https://github.com/hyattpd/Prodigal) suggested for Prokaryotes\n",
      "9. Quantification of genes identified in assemblies:\n",
      "   1. Generate index of assembly ([`BBmap index`](https://sourceforge.net/projects/bbmap/))\n",
      "   2. Mapping cleaned reads to the assembly for quantification ([`BBmap`](https://sourceforge.net/projects/bbmap/))\n",
      "   3. Get raw counts per each gene present in the assembly ([`Featurecounts`](http://subread.sourceforge.net)) -> TSV table with collected featurecounts output\n",
      "10. Functional annotation:\n",
      "    1. [`Eggnog`](https://github.com/eggnogdb/eggnog-mapper) -> Reformat TSV output \"eggnog table\"\n",
      "    2. [`KOfamscan`](https://github.com/takaram/kofam_scan)\n",
      "    3. [`HMMERsearch`](https://www.ebi.ac.uk/Tools/hmmer/search/hmmsearch) -> Ranking orfs based on HMMprofile with [`Hmmrank`](https://github.com/erikrikarddaniel/hmmrank)\n",
      "11. Taxonomic annotation:\n",
      "    1. [`EUKulele`](https://github.com/AlexanderLabWHOI/EUKulele) -> Reformat TSV output \"Reformat_tax.R\"\n",
      "    2. [`CAT`](https://github.com/dutilh/CAT)\n",
      "12. Summary statistics table. \"Collect_stats.R\"\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def retrieve(query: str, top_k: int) -> List[str]:\n",
    "    query_embedding = embed_chunk(query)\n",
    "    results = chromadb_collection.query(\n",
    "        query_embeddings=[query_embedding],\n",
    "        n_results=top_k,\n",
    "        include=[\"documents\", \"distances\"]  # 建议带上距离，方便观察\n",
    "    )\n",
    "    return results['documents'][0]\n",
    "\n",
    "# 测试\n",
    "query = \"What is metatdenovo?How to use it?\"\n",
    "retrieved_chunks = retrieve(query, 5)\n",
    "\n",
    "for i, chunk in enumerate(retrieved_chunks):\n",
    "    print(f\"[{i}] {chunk}\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a67bc819-692c-4474-8b2e-e96ae14271da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0] Introduction\n",
      "\n",
      "**nf-core/metatdenovo** is a bioinformatics best-practice analysis pipeline for assembly and annotation of metatranscriptomic and metagenomic data from prokaryotes, eukaryotes or viruses.\n",
      "\n",
      "On release, automated continuous integration tests run the pipeline on a full-sized dataset on the AWS cloud infrastructure. This ensures that the pipeline runs on AWS, has sensible resource allocation defaults set to run on real-world datasets, and permits the persistent storage of results to benchmark between pipeline releases and other analysis sources. The results obtained from the full-sized test can be viewed on the [nf-core website](https://nf-co.re/metatdenovo/results).\n",
      "\n",
      "\n",
      "\n",
      "[1] Usage\n",
      "\n",
      "![nf-core/metatdenovo metro map](docs/images/metat-metromap.png)\n",
      "\n",
      "1. Read QC ([`FastQC`](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/))\n",
      "2. Present QC for raw reads ([`MultiQC`](http://multiqc.info/))\n",
      "3. Quality trimming and adapter removal for raw reads ([`Trim Galore!`](https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/))\n",
      "4. Optional: Filter sequences with [`BBduk`](https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbduk-guide/)\n",
      "5. Optional: Normalize the sequencing depth with [`BBnorm`](https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbnorm-guide/)\n",
      "6. Merge trimmed, pair-end reads ([`Seqtk`](https://github.com/lh3/seqtk))\n",
      "7. Choice of de novo assembly programs:\n",
      "   1. [`RNAspades`](https://cab.spbu.ru/software/rnaspades/) suggested for Eukaryote de novo assembly\n",
      "   2. [`Megahit`](https://github.com/voutcn/megahit) suggested for Prokaryote de novo assembly\n",
      "8. Choice of orf caller:\n",
      "   1. [`TransDecoder`](https://github.com/TransDecoder/TransDecoder) suggested for Eukaryotes\n",
      "   2. [`Prokka`](https://github.com/tseemann/prokka) suggested for Prokaryotes\n",
      "   3. [`Prodigal`](https://github.com/hyattpd/Prodigal) suggested for Prokaryotes\n",
      "9. Quantification of genes identified in assemblies:\n",
      "   1. Generate index of assembly ([`BBmap index`](https://sourceforge.net/projects/bbmap/))\n",
      "   2. Mapping cleaned reads to the assembly for quantification ([`BBmap`](https://sourceforge.net/projects/bbmap/))\n",
      "   3. Get raw counts per each gene present in the assembly ([`Featurecounts`](http://subread.sourceforge.net)) -> TSV table with collected featurecounts output\n",
      "10. Functional annotation:\n",
      "    1. [`Eggnog`](https://github.com/eggnogdb/eggnog-mapper) -> Reformat TSV output \"eggnog table\"\n",
      "    2. [`KOfamscan`](https://github.com/takaram/kofam_scan)\n",
      "    3. [`HMMERsearch`](https://www.ebi.ac.uk/Tools/hmmer/search/hmmsearch) -> Ranking orfs based on HMMprofile with [`Hmmrank`](https://github.com/erikrikarddaniel/hmmrank)\n",
      "11. Taxonomic annotation:\n",
      "    1. [`EUKulele`](https://github.com/AlexanderLabWHOI/EUKulele) -> Reformat TSV output \"Reformat_tax.R\"\n",
      "    2. [`CAT`](https://github.com/dutilh/CAT)\n",
      "12. Summary statistics table. \"Collect_stats.R\"\n",
      "\n",
      "\n",
      "\n",
      "[2] Contributions and Support\n",
      "\n",
      "If you would like to contribute to this pipeline, please see the [contributing guidelines](.github/CONTRIBUTING.md).\n",
      "\n",
      "For further information or help, don't hesitate to get in touch on the [Slack `#metatdenovo` channel](https://nfcore.slack.com/channels/metatdenovo) (you can join with [this invite](https://nf-co.re/join/slack)).\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from sentence_transformers import CrossEncoder\n",
    "\n",
    "def rerank(query:str, retrieved_chunks:List[str],top_k:int) -> List[str]:\n",
    "    cross_encoder = CrossEncoder('cross-encoder/mmarco-mMiniLMv2-L12-H384-v1')\n",
    "    pairs=[(query,chunk) for chunk in retrieved_chunks]\n",
    "    scores=cross_encoder.predict(pairs)\n",
    "\n",
    "    chunk_with_score_list=[(chunk,score) \n",
    "                           for chunk,score in zip(retrieved_chunks, scores)]\n",
    "    chunk_with_score_list.sort(key=lambda pair:pair[1],reverse=True)\n",
    "    return [chunk for chunk,_ in chunk_with_score_list][:top_k]\n",
    "\n",
    "reranked_chunks = rerank(query, retrieved_chunks,3)\n",
    "\n",
    "for i,chunk in enumerate(reranked_chunks):\n",
    "    print(f\"[{i}]{chunk}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9078acaa-7c5c-4aad-a353-506b8e406723",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "You're a knowledge assistant,please answer questions according to the user's request and the following information. \n",
      "    the user's question:What is metatdenovo?How to use it?\n",
      "    relevant information:\n",
      "     Introduction\n",
      "\n",
      "**nf-core/metatdenovo** is a bioinformatics best-practice analysis pipeline for assembly and annotation of metatranscriptomic and metagenomic data from prokaryotes, eukaryotes or viruses.\n",
      "\n",
      "On release, automated continuous integration tests run the pipeline on a full-sized dataset on the AWS cloud infrastructure. This ensures that the pipeline runs on AWS, has sensible resource allocation defaults set to run on real-world datasets, and permits the persistent storage of results to benchmark between pipeline releases and other analysis sources. The results obtained from the full-sized test can be viewed on the [nf-core website](https://nf-co.re/metatdenovo/results).\n",
      "\n",
      "\n",
      "\n",
      " Usage\n",
      "\n",
      "![nf-core/metatdenovo metro map](docs/images/metat-metromap.png)\n",
      "\n",
      "1. Read QC ([`FastQC`](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/))\n",
      "2. Present QC for raw reads ([`MultiQC`](http://multiqc.info/))\n",
      "3. Quality trimming and adapter removal for raw reads ([`Trim Galore!`](https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/))\n",
      "4. Optional: Filter sequences with [`BBduk`](https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbduk-guide/)\n",
      "5. Optional: Normalize the sequencing depth with [`BBnorm`](https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbnorm-guide/)\n",
      "6. Merge trimmed, pair-end reads ([`Seqtk`](https://github.com/lh3/seqtk))\n",
      "7. Choice of de novo assembly programs:\n",
      "   1. [`RNAspades`](https://cab.spbu.ru/software/rnaspades/) suggested for Eukaryote de novo assembly\n",
      "   2. [`Megahit`](https://github.com/voutcn/megahit) suggested for Prokaryote de novo assembly\n",
      "8. Choice of orf caller:\n",
      "   1. [`TransDecoder`](https://github.com/TransDecoder/TransDecoder) suggested for Eukaryotes\n",
      "   2. [`Prokka`](https://github.com/tseemann/prokka) suggested for Prokaryotes\n",
      "   3. [`Prodigal`](https://github.com/hyattpd/Prodigal) suggested for Prokaryotes\n",
      "9. Quantification of genes identified in assemblies:\n",
      "   1. Generate index of assembly ([`BBmap index`](https://sourceforge.net/projects/bbmap/))\n",
      "   2. Mapping cleaned reads to the assembly for quantification ([`BBmap`](https://sourceforge.net/projects/bbmap/))\n",
      "   3. Get raw counts per each gene present in the assembly ([`Featurecounts`](http://subread.sourceforge.net)) -> TSV table with collected featurecounts output\n",
      "10. Functional annotation:\n",
      "    1. [`Eggnog`](https://github.com/eggnogdb/eggnog-mapper) -> Reformat TSV output \"eggnog table\"\n",
      "    2. [`KOfamscan`](https://github.com/takaram/kofam_scan)\n",
      "    3. [`HMMERsearch`](https://www.ebi.ac.uk/Tools/hmmer/search/hmmsearch) -> Ranking orfs based on HMMprofile with [`Hmmrank`](https://github.com/erikrikarddaniel/hmmrank)\n",
      "11. Taxonomic annotation:\n",
      "    1. [`EUKulele`](https://github.com/AlexanderLabWHOI/EUKulele) -> Reformat TSV output \"Reformat_tax.R\"\n",
      "    2. [`CAT`](https://github.com/dutilh/CAT)\n",
      "12. Summary statistics table. \"Collect_stats.R\"\n",
      "\n",
      "\n",
      "\n",
      " Contributions and Support\n",
      "\n",
      "If you would like to contribute to this pipeline, please see the [contributing guidelines](.github/CONTRIBUTING.md).\n",
      "\n",
      "For further information or help, don't hesitate to get in touch on the [Slack `#metatdenovo` channel](https://nfcore.slack.com/channels/metatdenovo) (you can join with [this invite](https://nf-co.re/join/slack)).\n",
      "\n",
      "\n",
      "    Please answer according to the information mentioned above.Do not make up information.\n",
      "\n",
      "---\n",
      "\n",
      "**What is metatdenovo?**\n",
      "nf-core/metatdenovo is a bioinformatics best-practice analysis pipeline designed for the assembly and annotation of metatranscriptomic and metagenomic data. It can be used for data originating from prokaryotes, eukaryotes, or viruses.\n",
      "\n",
      "**How to use it?**\n",
      "The nf-core/metatdenovo pipeline involves a series of steps for data processing and analysis:\n",
      "\n",
      "1.  **Read QC**: Initial quality control of raw reads using `FastQC`.\n",
      "2.  **Present QC for raw reads**: Summarizing raw read quality control with `MultiQC`.\n",
      "3.  **Quality trimming and adapter removal**: Cleaning raw reads using `Trim Galore!`.\n",
      "4.  **Optional Filtering**: Filtering sequences with `BBduk`.\n",
      "5.  **Optional Normalization**: Normalizing sequencing depth with `BBnorm`.\n",
      "6.  **Merge trimmed reads**: Merging trimmed, pair-end reads using `Seqtk`.\n",
      "7.  **De novo assembly**:\n",
      "    *   `RNAspades` is suggested for Eukaryote de novo assembly.\n",
      "    *   `Megahit` is suggested for Prokaryote de novo assembly.\n",
      "8.  **ORF calling**:\n",
      "    *   `TransDecoder` is suggested for Eukaryotes.\n",
      "    *   `Prokka` or `Prodigal` are suggested for Prokaryotes.\n",
      "9.  **Gene Quantification**:\n",
      "    *   Generating an index of the assembly using `BBmap index`.\n",
      "    *   Mapping cleaned reads to the assembly for quantification with `BBmap`.\n",
      "    *   Obtaining raw counts per gene in the assembly using `Featurecounts`.\n",
      "10. **Functional Annotation**:\n",
      "    *   `Eggnog`\n",
      "    *   `KOfamscan`\n",
      "    *   `HMMERsearch` with `Hmmrank` for ranking ORFs.\n",
      "11. **Taxonomic Annotation**:\n",
      "    *   `EUKulele`\n",
      "    *   `CAT`\n",
      "12. **Summary statistics table**: Collection of final statistics using \"Collect_stats.R\".\n"
     ]
    }
   ],
   "source": [
    "from dotenv import load_dotenv\n",
    "from google import genai\n",
    "\n",
    "\n",
    "load_dotenv()\n",
    "google_client = genai.Client()\n",
    "\n",
    "def generate(query:str, chunks:List[str]) ->str:\n",
    "    prompt = f\"\"\"You're a knowledge assistant,please answer questions according to the user's request and the following information. \n",
    "    the user's question:{query}\n",
    "    relevant information:\n",
    "    {\"\\n\\n\".join(chunks)}\n",
    "    Please answer according to the information mentioned above.Do not make up information.\"\"\"\n",
    "\n",
    "    print(f\"{prompt}\\n\\n---\\n\")\n",
    "\n",
    "    response = google_client.models.generate_content(\n",
    "        model=\"gemini-2.5-flash\",\n",
    "        contents=prompt\n",
    "    )\n",
    "    return response.text\n",
    "\n",
    "answer = generate(query,reranked_chunks)\n",
    "print(answer)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "896fc29e-a5d2-4e21-a3d7-efa8f89d5eed",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
